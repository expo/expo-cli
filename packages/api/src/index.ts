// Analytics
export { default as Analytics } from './Analytics';
export { default as UnifiedAnalytics } from './UnifiedAnalytics';
// Utilities
export * as ExpoConfigSchema from './ExpoConfigSchema';
export { Cache } from './Cache';

// Underlying API
export { default as ApiV2 } from './ApiV2';

// API
export * as StandaloneBuild from './StandaloneBuild';
export * as Publish from './Publish';
export * as Assets from './Assets';
export * as Manifest from './Manifest';
export * as DevelopmentSessions from './DevelopmentSessions';
export * as Sdks from './Sdks';
export * as Projects from './Projects';
export * as SendProject from './SendProject';
export * as Auth from './Auth';
export * as Versions from './Versions';

// Non-standard API (TODO)
export { default as UserManager } from './UserManager';

// Settings
export { default as ProcessSettings } from './ProcessSettings';
export { default as UserSettings } from './UserSettings';
export * as ProjectSettings from './ProjectSettings';
