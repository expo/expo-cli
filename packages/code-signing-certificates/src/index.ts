import assert from 'assert';
import { md, pki as PKI, random, util } from 'node-forge';

import { toPositiveHex } from './utils';

/**
 * Generate a public and private RSA key pair.
 * @returns RSA key pair
 */
export function generateKeyPair(): PKI.rsa.KeyPair {
  return PKI.rsa.generateKeyPair();
}

/**
 * Convert a key RSA key pair generated using {@link generateKeyPair} to PEM strings.
 * @param keyPair RSA key pair
 * @returns PEM formatted key pair
 */
export function convertKeyPairToPEM(
  keyPair: PKI.rsa.KeyPair
): { privateKeyPEM: string; publicKeyPEM: string } {
  return {
    privateKeyPEM: PKI.privateKeyToPem(keyPair.privateKey),
    publicKeyPEM: PKI.publicKeyToPem(keyPair.publicKey),
  };
}

/**
 * Convert a X.509 certificate generated using {@link generateSelfSignedCodeSigningCertificate} to a PEM string.
 * @param certificate X.509 certificate
 * @returns
 */
export function convertCertificateToCertificatePEM(certificate: PKI.Certificate): string {
  return PKI.certificateToPem(certificate);
}

/**
 * Convert a PEM-formatted RSA key pair to a key pair for use with this library.
 * @param privateKeyPEM PEM formatted private key
 * @param publicKeyPEM PEM formatted public key
 * @returns RSA key pair
 */
export function convertKeyPairPEMToKeyPair({
  privateKeyPEM,
  publicKeyPEM,
}: {
  privateKeyPEM: string;
  publicKeyPEM: string;
}): PKI.rsa.KeyPair {
  return {
    privateKey: PKI.privateKeyFromPem(privateKeyPEM),
    publicKey: PKI.publicKeyFromPem(publicKeyPEM),
  };
}

/**
 * Convert a PEM-formatted RSA public key to a public key for use with this library.
 * @param publicKeyPEM PEM formatted public key
 * @returns RSA public key
 */
export function convertPublicKeyPEMToPublicKey(publicKeyPEM: string): PKI.rsa.PublicKey {
  return PKI.publicKeyFromPem(publicKeyPEM);
}

/**
 * Convert a PEM-formatted RSA private key to a private key for use with this library.
 * @param privateKeyPEM PEM formatted private key
 * @returns RSA private key
 */
export function convertPrivateKeyPEMToPrivateKey(privateKeyPEM: string): PKI.rsa.PrivateKey {
  return PKI.privateKeyFromPem(privateKeyPEM);
}

/**
 * Convert a PEM-formatted X.509 certificate to a certificate for use with this library.
 * @param certificatePEM PEM formatted X.509 certificate
 * @returns  X.509 Certificate
 */
export function convertCertificatePEMToCertificate(certificatePEM: string): PKI.Certificate {
  return PKI.certificateFromPem(certificatePEM);
}

type GenerateParameters = {
  /**
   * Public/private key pair generated via {@link generateKeyPair}.
   */
  keyPair: PKI.rsa.KeyPair;

  /**
   * Certificate validity range start.
   */
  validityNotBefore: Date;

  /**
   * Certificate validity range end.
   */
  validityNotAfter: Date;

  /**
   * CN issuer and subject Distinguished Name (DN).
   * Used for both issuer and subject in the case of self-signed certificates.
   */
  commonName: string;
};

/**
 * Generate a self-signed code-signing certificate for use with expo-updates.
 * Note that while certificate chains may be supported at some point in expo-updates, for now
 * only self-signed certificates are supported.
 *
 * @returns PKI.Certificate valid for expo-updates code signing
 */
export function generateSelfSignedCodeSigningCertificate({
  keyPair: { publicKey, privateKey },
  validityNotBefore,
  validityNotAfter,
  commonName,
}: GenerateParameters): PKI.Certificate {
  const cert = PKI.createCertificate();
  cert.publicKey = publicKey;
  cert.serialNumber = toPositiveHex(util.bytesToHex(random.getBytesSync(9)));

  assert(
    validityNotAfter > validityNotBefore,
    'validityNotAfter must be later than validityNotBefore'
  );
  cert.validity.notBefore = validityNotBefore;
  cert.validity.notAfter = validityNotAfter;

  const attrs = [
    {
      name: 'commonName',
      value: commonName,
    },
  ];
  cert.setSubject(attrs);
  cert.setIssuer(attrs);

  cert.setExtensions([
    {
      name: 'keyUsage',
      critical: true,
      keyCertSign: false,
      digitalSignature: true,
      nonRepudiation: false,
      keyEncipherment: false,
      dataEncipherment: false,
    },
    {
      name: 'extKeyUsage',
      critical: true,
      serverAuth: false,
      clientAuth: false,
      codeSigning: true,
      emailProtection: false,
      timeStamping: false,
    },
  ]);

  cert.sign(privateKey, md.sha256.create());
  return cert;
}

function arePublicKeysEqual(key1: PKI.rsa.PublicKey, key2: PKI.rsa.PublicKey): boolean {
  // typedef for BigInteger doesn't contain equals method
  return (key1.n as any).equals(key2.n) && (key1.e as any).equals(key2.e);
}

function doPrivateAndPublicKeysMatch(
  privateKey: PKI.rsa.PrivateKey,
  publicKey: PKI.rsa.PublicKey
): boolean {
  // typedef for BigInteger doesn't contain equals method
  return (publicKey.n as any).equals(privateKey.n) && (publicKey.e as any).equals(privateKey.e);
}

/**
 * Validate that a certificate and corresponding key pair can be used for expo-updates code signing.
 * @param certificate X.509 certificate
 * @param keyPair RSA key pair
 */
export function validateSelfSignedCertificate(
  certificate: PKI.Certificate,
  keyPair: PKI.rsa.KeyPair
) {
  if (certificate.issuer.hash !== certificate.subject.hash) {
    throw new Error(
      'Certificate issuer hash does not match subject hash, indicating certificate is not self-signed.'
    );
  }

  const now = new Date();
  if (certificate.validity.notBefore > now || certificate.validity.notAfter < now) {
    throw new Error('Certificate validity expired');
  }

  const keyUsage = certificate.getExtension('keyUsage');
  const digitalSignature = (keyUsage as any).digitalSignature;
  if (!keyUsage || !digitalSignature) {
    throw new Error('X509v3 Key Usage: Digital Signature not present');
  }

  const extKeyUsage = certificate.getExtension('extKeyUsage');
  const codeSigning = (extKeyUsage as any).codeSigning;
  if (!extKeyUsage || !codeSigning) {
    throw new Error('X509v3 Extended Key Usage: Code Signing not present');
  }

  const isValid = certificate.verify(certificate);
  if (!isValid) {
    throw new Error('Certificate signature not valid');
  }

  const certificatePublicKey = certificate.publicKey as PKI.rsa.PublicKey;
  if (!arePublicKeysEqual(certificatePublicKey, keyPair.publicKey)) {
    throw new Error('Certificate pubic key does not match key pair public key');
  }

  if (!doPrivateAndPublicKeysMatch(keyPair.privateKey, keyPair.publicKey)) {
    throw new Error('keyPair key mismatch');
  }
}

/**
 * Sign a string with an RSA private key and verify that the signature is valid for the RSA
 * public key in the certificate.
 * @param privateKey RSA private key
 * @param certificate X.509 certificate
 * @param stringToSign string for which to generate a signature and verify
 * @returns base64-encoded signature
 */
export function signStringRSASHA256AndVerify(
  privateKey: PKI.rsa.PrivateKey,
  certificate: PKI.Certificate,
  stringToSign: string
): string {
  const digest = md.sha256.create().update(stringToSign);
  const digestSignature = privateKey.sign(digest);
  const isValidSignature = (certificate.publicKey as PKI.rsa.PublicKey).verify(
    digest.digest().getBytes(),
    digestSignature
  );

  if (!isValidSignature) {
    throw new Error('Signature generated with private key not valid for certificate');
  }

  return util.encode64(digestSignature);
}
