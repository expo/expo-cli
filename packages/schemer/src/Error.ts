import ExtendableError from 'es6-error';

export class SchemerError extends ExtendableError {
  errors: Array<ValidationError>;
  constructor(errors: Array<ValidationError>) {
    const message = errors.map(e => e.message).join('\n');
    super(message);
    this.errors = errors;
  }
}

export class ValidationError extends ExtendableError {
  errorCode: string;
  fieldPath: string;
  message: string;
  data: any;
  meta: any;
  constructor({
    errorCode,
    fieldPath,
    message,
    data,
    meta,
  }: {
    errorCode: ErrorCode;
    fieldPath: string;
    message: string;
    data: any;
    meta: any;
  }) {
    super(message);
    this.errorCode = errorCode;
    this.fieldPath = fieldPath;
    this.message = message;
    this.data = data;
    this.meta = meta;
  }
}

export type ErrorCode = keyof typeof ErrorCodes;

export const ErrorCodes = {
  SCHEMA_VALIDATION_ERROR: 'SCHEMA_VALIDATION_ERROR',
  SCHEMA_ADDITIONAL_PROPERTY: 'SCHEMA_ADDITIONAL_PROPERTY',
  SCHEMA_MISSING_REQUIRED_PROPERTY: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
  SCHEMA_INVALID_PATTERN: 'SCHEMA_INVALID_PATTERN',

  INVALID_ASSET_URI: 'INVALID_ASSET_URI',
  INVALID_DIMENSIONS: 'INVALID_DIMENSIONS',
  INVALID_CONTENT_TYPE: 'INVALID_CONTENT_TYPE',
  NOT_SQUARE: 'NOT_SQUARE',
};
