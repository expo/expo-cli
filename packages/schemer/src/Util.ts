import fill from 'lodash/fill';
import flatten from 'lodash/flatten';
import get from 'lodash/get';
import zip from 'lodash/zip';

export const fieldPathToSchemaPath = (fieldPath: string) => {
  let newPath = zip(fill(fieldPath.split('.'), 'properties'), fieldPath.split('.'));
  return flatten(newPath).join('.');
};
// Assumption: used only for jsonPointer returned from traverse
export const schemaPointerToFieldPath = (jsonPointer: string) => {
  return jsonPointer
    .split('/')
    .slice(2)
    .filter(e => e !== 'properties')
    .join('.');
};

export const fieldPathToSchema = (schema: Object, fieldPath: string) => {
  return get(schema, fieldPathToSchemaPath(fieldPath));
};
