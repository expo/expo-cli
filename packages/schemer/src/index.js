/**
 * @flow
 **/

import 'babel-polyfill';
import _ from 'lodash';
import Ajv from 'ajv';
import path from 'path';
import fs from 'fs';
import traverse from 'json-schema-traverse';
import readChunk from 'read-chunk';
import imageProbe from 'probe-image-size';
import { SchemerError, ValidationError, ErrorCodes } from './Error';
import { schemaPointerToFieldPath, fieldPathToSchema } from './Util';

type Options = {
  allErrors?: boolean,
  rootDir?: string,
  verbose?: boolean,
  format?: 'full' | 'empty', //figure out later
  metaValidation?: boolean,
};

type Meta = {
  asset?: boolean,
  dimensions?: Object,
  square: boolean,
  contentTypePattern: string,
  contentTypeHuman: string,
};

export { SchemerError, ValidationError, ErrorCodes } from './Error';
export default class Schemer {
  options: Options;
  ajv: Object;
  schema: Object;
  rootDir: string;
  manualValidationErrors: Array<ValidationError>;
  // Schema is a JSON Schema object
  constructor(schema: Object, options: Options = {}) {
    this.options = _.extend(
      {
        allErrors: true,
        verbose: true,
        format: 'full',
        metaValidation: true,
      },
      options
    );

    this.ajv = new Ajv(this.options);
    this.schema = schema;
    this.rootDir = this.options.rootDir || __dirname;
    this.manualValidationErrors = [];
  }

  _formatAjvErrorMessage({
    keyword,
    dataPath,
    params,
    parentSchema,
    data,
    message,
  }: {
    keyword: string,
    dataPath: string,
    params: Object,
    parentSchema: Object,
    data: any,
    message: string,
  }) {
    // This removes the "." in front of a fieldPath
    dataPath = dataPath.slice(1);
    switch (keyword) {
      case 'additionalProperties':
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_ADDITIONAL_PROPERTY,
          fieldPath: dataPath,
          message: `should NOT have additional property '${params.additionalProperty}'`,
          data,
          meta: parentSchema.meta,
        });
      case 'required':
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_MISSING_REQUIRED_PROPERTY,
          fieldPath: dataPath,
          message: `is missing required property '${params.missingProperty}'`,
          data,
          meta: parentSchema.meta,
        });
      case 'pattern': {
        //@TODO Parse the message in a less hacky way. Perhaps for regex validation errors, embed the error message under the meta tag?
        const regexHuman = _.get(parentSchema, 'meta.regexHuman');
        const regexErrorMessage = regexHuman
          ? `'${dataPath}' should be a ${regexHuman[0].toLowerCase() + regexHuman.slice(1)}`
          : `'${dataPath}' ${message}`;
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_INVALID_PATTERN,
          fieldPath: dataPath,
          message: regexErrorMessage,
          data,
          meta: parentSchema.meta,
        });
      }
      default:
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_VALIDATION_ERROR,
          fieldPath: dataPath,
          message,
          data,
          meta: parentSchema.meta,
        });
    }
  }

  getErrors(): Array<ValidationError> {
    // Convert AJV JSONSchema errors to our ValidationErrors
    let valErrors = [];
    if (this.ajv.errors) {
      valErrors = this.ajv.errors.map(e => this._formatAjvErrorMessage(e));
    }
    const bothErrors = _.concat(valErrors, this.manualValidationErrors);
    return bothErrors;
  }

  _throwOnErrors() {
    // Clean error state after each validation
    const errors = this.getErrors();
    if (errors.length > 0) {
      this.manualValidationErrors = [];
      this.ajv.errors = [];
      throw new SchemerError(errors);
    }
  }

  async validateAll(data: any) {
    await this._validateSchemaAsync(data);
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateAssetsAsync(data: any) {
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateSchemaAsync(data: any) {
    await this._validateSchemaAsync(data);
    this._throwOnErrors();
  }

  _validateSchemaAsync(data: any) {
    this.ajv.validate(this.schema, data);
  }

  async _validateAssetsAsync(data: any) {
    let assets = [];
    traverse(this.schema, { allKeys: true }, (subSchema, jsonPointer, a, b, c, d, property) => {
      if (property && subSchema.meta && subSchema.meta.asset) {
        const fieldPath = schemaPointerToFieldPath(jsonPointer);
        assets.push({
          fieldPath,
          data: _.get(data, fieldPath),
          meta: subSchema.meta,
        });
      }
    });
    await Promise.all(assets.map(this._validateAssetAsync.bind(this)));
  }

  async _validateAssetAsync({
    fieldPath,
    data,
    meta,
  }: {
    fieldPath: string,
    data: string,
    meta: Meta,
  }) {
    if (meta && meta.asset && data) {
      const { asset, dimensions, square, contentTypePattern, contentTypeHuman }: Meta = meta;
      // filePath could be an URL
      const filePath = path.resolve(this.rootDir, data);
      try {
        // Assumption: All assets are images. This may change in the future.
        // NOTE(nikki): The '4100' below should be enough for most file types, though we
        //              could probably go shorter....
        //              http://www.garykessler.net/library/file_sigs.html
        //  The metadata content for .jpgs might be located a lot farther down the file, so this
        //  may pose problems in the future.
        //  This cases on whether filePath is a remote URL or located on the machine
        const probeResult = fs.existsSync(filePath)
          ? imageProbe.sync(await readChunk(filePath, 0, 4100))
          : await imageProbe(data, { useElectronNet: false });

        const { width, height, type, mime, wUnits, hUnits } = probeResult;

        if (contentTypePattern && !mime.match(new RegExp(contentTypePattern))) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.INVALID_CONTENT_TYPE,
              fieldPath,
              message: `field '${fieldPath}' should point to ${
                meta.contentTypeHuman
              } but the file at '${data}' has type ${type}`,
              data,
              meta,
            })
          );
        }

        if (dimensions && (dimensions.height !== height || dimensions.width != width)) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.INVALID_DIMENSIONS,
              fieldPath,
              message: `'${fieldPath}' should have dimensions ${dimensions.width}x${
                dimensions.height
              }, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }

        if (square && width !== height) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.NOT_SQUARE,
              fieldPath,
              message: `image should be square, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }
      } catch (e) {
        this.manualValidationErrors.push(
          new ValidationError({
            errorCode: ErrorCodes.INVALID_ASSET_URI,
            fieldPath,
            message: `cannot access file at '${data}'`,
            data,
            meta,
          })
        );
      }
    }
  }

  async validateProperty(fieldPath: string, data: any) {
    const subSchema = fieldPathToSchema(this.schema, fieldPath);
    this.ajv.validate(subSchema, data);

    if (subSchema.meta && subSchema.meta.asset) {
      await this._validateAssetAsync({ fieldPath, data, meta: subSchema.meta });
    }
    this._throwOnErrors();
  }

  validateName(name: string) {
    return this.validateProperty('name', name);
  }

  validateSlug(slug: string) {
    return this.validateProperty('slug', slug);
  }

  validateSdkVersion(version: string) {
    return this.validateProperty('sdkVersion', version);
  }

  validateIcon(iconPath: string) {
    return this.validateProperty('icon', iconPath);
  }
}
