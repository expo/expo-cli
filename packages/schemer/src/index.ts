import path from 'path';
import fs from 'fs';

import _ from 'lodash';
import Ajv from 'ajv';
import imageProbe from 'probe-image-size';
import readChunk from 'read-chunk';
import traverse from 'json-schema-traverse';

import { schemaPointerToFieldPath, fieldPathToSchema } from './Util';
import { SchemerError, ValidationError } from './Error';

type Options = {
  allErrors?: boolean;
  rootDir?: string;
  verbose?: boolean;
  format?: 'full' | 'empty'; //figure out later
  metaValidation?: boolean;
};

type Meta = {
  asset?: boolean;
  dimensions?: {
    width: number;
    height: number;
  };
  square?: boolean;
  contentTypePattern?: string;
  contentTypeHuman?: string;
};

type AssetField = { fieldPath: string; data: string; meta: Meta };

export { SchemerError, ValidationError, ErrorCodes, ErrorCode } from './Error';
export default class Schemer {
  options: Options;
  ajv: Ajv.Ajv;
  schema: Object;
  rootDir: string;
  manualValidationErrors: Array<ValidationError>;
  // Schema is a JSON Schema object
  constructor(schema: Object, options: Options = {}) {
    this.options = _.extend(
      {
        allErrors: true,
        verbose: true,
        format: 'full',
        metaValidation: true,
      },
      options
    );

    this.ajv = new Ajv(this.options);
    this.schema = schema;
    this.rootDir = this.options.rootDir || __dirname;
    this.manualValidationErrors = [];
  }

  _formatAjvErrorMessage({
    keyword,
    dataPath,
    params,
    parentSchema,
    data,
    message,
  }: Ajv.ErrorObject) {
    const meta = parentSchema && (parentSchema as any).meta;
    // This removes the "." in front of a fieldPath
    dataPath = dataPath.slice(1);
    switch (keyword) {
      case 'additionalProperties': {
        return new ValidationError({
          errorCode: 'SCHEMA_ADDITIONAL_PROPERTY',
          fieldPath: dataPath,
          message: `should NOT have additional property '${(params as any).additionalProperty}'`,
          data,
          meta,
        });
      }
      case 'required':
        return new ValidationError({
          errorCode: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
          fieldPath: dataPath,
          message: `is missing required property '${(params as any).missingProperty}'`,
          data,
          meta,
        });
      case 'pattern': {
        //@TODO Parse the message in a less hacky way. Perhaps for regex validation errors, embed the error message under the meta tag?
        const regexHuman = _.get(parentSchema, 'meta.regexHuman');
        const regexErrorMessage = regexHuman
          ? `'${dataPath}' should be a ${regexHuman[0].toLowerCase() + regexHuman.slice(1)}`
          : `'${dataPath}' ${message}`;
        return new ValidationError({
          errorCode: 'SCHEMA_INVALID_PATTERN',
          fieldPath: dataPath,
          message: regexErrorMessage,
          data,
          meta,
        });
      }
      default:
        return new ValidationError({
          errorCode: 'SCHEMA_VALIDATION_ERROR',
          fieldPath: dataPath,
          message: message || 'Validation error',
          data,
          meta,
        });
    }
  }

  getErrors(): ValidationError[] {
    // Convert AJV JSONSchema errors to our ValidationErrors
    let valErrors: ValidationError[] = [];
    if (this.ajv.errors) {
      valErrors = this.ajv.errors.map(e => this._formatAjvErrorMessage(e));
    }
    const bothErrors = _.concat(valErrors, this.manualValidationErrors);
    return bothErrors;
  }

  _throwOnErrors() {
    // Clean error state after each validation
    const errors = this.getErrors();
    if (errors.length > 0) {
      this.manualValidationErrors = [];
      this.ajv.errors = [];
      throw new SchemerError(errors);
    }
  }

  async validateAll(data: any) {
    await this._validateSchemaAsync(data);
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateAssetsAsync(data: any) {
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateSchemaAsync(data: any) {
    await this._validateSchemaAsync(data);
    this._throwOnErrors();
  }

  _validateSchemaAsync(data: any) {
    this.ajv.validate(this.schema, data);
  }

  async _validateAssetsAsync(data: any) {
    let assets: AssetField[] = [];
    traverse(this.schema, { allKeys: true }, (subSchema, jsonPointer, a, b, c, d, property) => {
      if (property && subSchema.meta && subSchema.meta.asset) {
        const fieldPath = schemaPointerToFieldPath(jsonPointer);
        assets.push({
          fieldPath,
          data: _.get(data, fieldPath),
          meta: subSchema.meta,
        });
      }
    });
    await Promise.all(assets.map(this._validateAssetAsync.bind(this)));
  }

  async _validateImageAsync({ fieldPath, data, meta }: AssetField) {
    if (meta && meta.asset && data) {
      const { dimensions, square, contentTypePattern }: Meta = meta;
      // filePath could be an URL
      const filePath = path.resolve(this.rootDir, data);
      try {
        // NOTE(nikki): The '4100' below should be enough for most file types, though we
        //              could probably go shorter....
        //              http://www.garykessler.net/library/file_sigs.html
        //  The metadata content for .jpgs might be located a lot farther down the file, so this
        //  may pose problems in the future.
        //  This cases on whether filePath is a remote URL or located on the machine
        const probeResult = fs.existsSync(filePath)
          ? imageProbe.sync(await readChunk(filePath, 0, 4100))
          : await imageProbe(data, { useElectronNet: false });
        if (!probeResult) {
          return;
        }

        const { width, height, type, mime } = probeResult;

        if (contentTypePattern && !mime.match(new RegExp(contentTypePattern))) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'INVALID_CONTENT_TYPE',
              fieldPath,
              message: `field '${fieldPath}' should point to ${
                meta.contentTypeHuman
              } but the file at '${data}' has type ${type}`,
              data,
              meta,
            })
          );
        }

        if (dimensions && (dimensions.height !== height || dimensions.width !== width)) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'INVALID_DIMENSIONS',
              fieldPath,
              message: `'${fieldPath}' should have dimensions ${dimensions.width}x${
                dimensions.height
              }, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }

        if (square && width !== height) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'NOT_SQUARE',
              fieldPath,
              message: `image should be square, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }
      } catch (e) {
        this.manualValidationErrors.push(
          new ValidationError({
            errorCode: 'INVALID_ASSET_URI',
            fieldPath,
            message: `cannot access file at '${data}'`,
            data,
            meta,
          })
        );
      }
    }
  }

  async _validateAssetAsync({ fieldPath, data, meta }: AssetField) {
    if (meta && meta.asset && data) {
      if (meta.contentTypePattern && meta.contentTypePattern.startsWith('^image')) {
        await this._validateImageAsync({ fieldPath, data, meta });
      }
    }
  }

  async validateProperty(fieldPath: string, data: any) {
    const subSchema = fieldPathToSchema(this.schema, fieldPath);
    this.ajv.validate(subSchema, data);

    if (subSchema.meta && subSchema.meta.asset) {
      await this._validateAssetAsync({ fieldPath, data, meta: subSchema.meta });
    }
    this._throwOnErrors();
  }

  validateName(name: string) {
    return this.validateProperty('name', name);
  }

  validateSlug(slug: string) {
    return this.validateProperty('slug', slug);
  }

  validateSdkVersion(version: string) {
    return this.validateProperty('sdkVersion', version);
  }

  validateIcon(iconPath: string) {
    return this.validateProperty('icon', iconPath);
  }
}
