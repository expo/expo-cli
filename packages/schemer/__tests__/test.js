import Schemer from '../src/index';
import { pathToSchema, pathToSchemaPath, schemaPointerToPath } from '../src/Util';
import { SchemerError, ValidationError, ErrorCodes } from '../src/Error.js';

describe('Sanity Tests', () => {
  it('is a class', () => {
    const schema = require('./files/schema.json');
    const S = new Schemer(schema);
    expect(S instanceof Schemer).toBe(true);
  });

  it('has public functions', () => {
    const schema = require('./files/schema.json');
    const S = new Schemer(schema);
    expect(S.validateAll).toBeDefined();
    expect(S.validateProperty).toBeDefined();
  });
});

const schema = require('./files/schema.json').schema;
const S = new Schemer(schema, { rootDir: './__tests__' });
const good = require('./files/app.json');
const bad = require('./files/bad.json');

describe('Holistic Unit Test', () => {
  it('good example app.json all', async () => {
    await expect(S.validateAll(good)).resolves;
  });

  it('good example app.json schema', async () => {
    await expect(S.validateSchemaAsync(good)).resolves;
  });

  it('bad example app.json schema', async () => {
    try {
      await S.validateSchemaAsync(bad);
    } catch (e) {
      console.log(e);
      expect(e).toBeTruthy();
      expect(e.errors.length).toBe(5);
    }
  });
});

describe('Manual Validation Individual Unit Tests', () => {
  it('Local Icon', async () => {
    await expect(S.validateIcon('./files/check.png')).resolves;
  });

  it('Local Square Icon correct', async () => {
    const S = new Schemer(
      { properties: { icon: { meta: { asset: true, square: true } } } },
      { rootDir: './__tests__' }
    );
    await expect(S.validateIcon('./files/check.png')).resolves;
  });

  it('Local icon dimensions wrong', async () => {
    const S = new Schemer({
      properties: {
        icon: {
          meta: { asset: true, dimensions: { width: 400, height: 401 } },
        },
      },
    });
    try {
      await S.validateIcon('./files/check.png');
    } catch (e) {
      expect(e).toBeTruthy();
      expect(e.errors.length).toBe(1);
    }
  });
});

describe('Individual Unit Tests', async () => {
  it('Error when missing Required Property', async () => {
    let S = new Schemer({
      properties: {
        name: {},
      },
      required: ['name'],
    });
    try {
      await S.validateAll({ noName: '' });
    } catch (e) {
      expect(e.errors.length).toBe(1);
      expect(e.errors[0].errorCode).toBe(ErrorCodes.SCHEMA_MISSING_REQUIRED_PROPERTY);
    }
  });

  it('Error when data has an additional property', async () => {
    let S = new Schemer({ additionalProperties: false });
    try {
      await S.validateAll({ extraProperty: 'extra' });
    } catch (e) {
      expect(e.errors.length).toBe(1);
      expect(e.errors[0].errorCode).toBe(ErrorCodes.SCHEMA_ADDITIONAL_PROPERTY);
    }
  });

  it('Name', async () => {
    await expect(S.validateName('wilson')).resolves;
    await expect(S.validateName([1, 2, 3, 4])).rejects.toBeDefined();
    await expect(S.validateName(23.232332)).rejects.toBeDefined();
    await expect(S.validateName(/regex.*/)).rejects.toBeDefined();
  });

  xit('Slug', async () => {
    await expect(S.validateSlug('wilson')).resolves;
    await expect(S.validateSlug(12312123123)).rejects.toBeDefined();
    await expect(S.validateSlug([1, 23])).rejects.toBeDefined();

    await expect(S.validateSlug('wilson123')).resolves;
    await expect(S.validateSlug('wilson-123')).resolves;
    await expect(S.validateSlug('wilson/test')).rejects.toBeDefined();
    await expect(S.validateSlug('wilson-test%')).rejects.toBeDefined();
    await expect(S.validateSlug('wilson-test-zhao--javascript-is-super-funky')).resolves;
  });

  xit('SDK Version', async () => {
    await expect(S.validateSdkVersion('1.0.0')).resolves;
    // TODO: is the following allowed?
    await expect(S.validateSdkVersion('2.0.0.0.1')).rejects.toBeDefined();
    await expect(S.validateSdkVersion('UNVERSIONED')).resolves;
    await expect(S.validateSdkVersion('12.2a.3')).rejects.toBeDefined();
    await expect(S.validateSdkVersion('9,9,9')).rejects.toBeDefined();
    await expect(S.validateSdkVersion('1.2')).rejects.toBeDefined();
  });
});
