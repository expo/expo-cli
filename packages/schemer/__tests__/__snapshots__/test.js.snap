// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holistic Unit Test bad example app.json schema 1`] = `
Array [
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_MISSING_REQUIRED_PROPERTY",
    "fieldPath": "",
    "message": "is missing required property 'name'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'orientaasdfasdftion'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'asdfasdfandroid'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": "1*@)#($*@)(#$*)",
    "errorCode": "SCHEMA_INVALID_PATTERN",
    "fieldPath": "slug",
    "message": "'slug' must match pattern \\"^[a-zA-Z0-9_\\\\-]+$\\"",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Holistic Unit Test bad example app.json schema with field with not 1`] = `
Array [
  Object {
    "data": "1.0",
    "errorCode": "SCHEMA_INVALID_NOT",
    "fieldPath": "runtimeVersion",
    "message": "'runtimeVersion' should be not a decimal ending in a 0.",
    "meta": Object {
      "notHuman": "Not a decimal ending in a 0.",
    },
    "name": "ValidationError",
  },
]
`;
